### ISSUE TITLE
Change UTF32 code point output to a UTF8

### ISSUE THREAD

**Issue by genusistimelord** [glfw/glfw#837]
There are a few functions on windows for their IME to give you a UTF16 string and then you could use a small converter to output UTF8. Other libraries like SDL already support this kind of output through their char callbacks. For example you would change the IME output to the character callback code so it outputs a char * instead of a UTF32 code point. A lot of which is based on: http://utf8everywhere.org/

typedef void (* GLFWcharfun)(GLFWwindow*, unsigned int); 
should be
typedef void (* GLFWcharfun)(GLFWwindow*, char *);

typedef void (* GLFWcharmodsfun)(GLFWwindow*, unsigned int, int);
should be
typedef void (* GLFWcharmodsfun)(GLFWwindow*, char *, int);

The reason why this should be done is so the strings work more effectively with default string handling functions. As by design all string functions can use and handle UTF8 and only require you to check for a UTF8 character if your removing a character or truncating the string. Also this process would speed up the returned string handling for others projects and speed up the utf8 creation on x11 or other systems that fully support utf8 returns. 

This is mainly a feature change and is solely up to you if you would want to make 3.4 or another version handle proper Unicode strings.

**Comment by elmindreda** [comment]
I agree, and this is already planned for 4.0. It can't happen during 3.x as it changes the API.

**Comment by milasudril** [comment]
Would it be possible to keep the current callback. Otherwise, I guess I would have the situation codepoint->UTF8 ->codepoint for character input. Or, does GLFW extract the codepoint from UTF8 already? It would add a new pair of functions:

    typedef void (_ GLFWutf8stringfun)(GLFWwindow*,const char *);
    typedef void (_ GLFWutf8stringfun)(GLFWwindow*,const char *, int);

This does not break existing API, so it can be done in version 3. Notice that I have put `const` in front of the pointer, because I think this is what you want.

**Comment by cblc** [comment]
Wasn't this addressed in the UTF32 to UTF8 conversion function added to `x11_window.c` by https://github.com/glfw/glfw/pull/983 ? I mean, why not keep the callbacks as they are now, and just make that function public, so that GLFW provides a function that converts from UTF32 to UTF8? Then, if your app expects UTF8 (which most likely will because most code works with UTF8), you call the UTF32 to UTF8 converter from your `character_callback`

**Comment by genusistimelord** [comment]
I would say yes that would solve the issue if it is accessible publicly and viewable on different Operating systems GLFW supports.
